File: Terraform/lambda_functions/clock_in_out.py
import json
import os
import boto3
from datetime import datetime
from decimal import Decimal
from botocore.exceptions import ClientError
from boto3.dynamodb.conditions import Key
from auth_utils import get_cors_headers, handle_cors_preflight, get_user_from_token

def lambda_handler(event, context):
    headers = get_cors_headers()
    
    # Handle preflight OPTIONS request
    cors_response = handle_cors_preflight(event)
    if cors_response:
        return cors_response

    try:
        # Parse request body
        body = json.loads(event['body'])
        action = body.get('action')  # 'in' or 'out'
        
        # Get and validate user from token
        user_id, auth_error = get_user_from_token(event)
        if auth_error:
            return auth_error

        dynamodb = boto3.resource('dynamodb')
        time_entries_table = dynamodb.Table(os.environ['TIME_ENTRIES_TABLE'])
        user_sessions_table = dynamodb.Table(os.environ['USER_SESSIONS_TABLE'])

        timestamp = datetime.utcnow().isoformat() + 'Z'
        date = timestamp.split('T')[0]

        if action == 'in':
            # Check if user is already clocked in
            try:
                session_response = user_sessions_table.get_item(
                    Key={'user_id': user_id}
                )
                session_item = session_response.get('Item')
                
                if session_item and session_item.get('status') == 'clocked_in':
                    return {
                        'statusCode': 400,
                        'headers': headers,
                        'body': json.dumps({'error': 'Already clocked in. Please clock out first.'})
                    }
            except ClientError as e:
                print(f"DynamoDB error checking session: {e}")
                # Continue with clock in if we can't check session status
            
            # Clock in
            time_entries_table.put_item(
                Item={
                    'user_id': user_id,
                    'timestamp': timestamp,
                    'date': date,
                    'type': 'clock_in',
                    'hours': Decimal('0')
                }
            )

            # Update user session
            user_sessions_table.put_item(
                Item={
                    'user_id': user_id,
                    'status': 'clocked_in',
                    'clock_in_time': timestamp,
                    'last_updated': timestamp
                }
            )

            return {
                'statusCode': 200,
                'headers': headers,
                'body': json.dumps({
                    'success': True,
                    'message': 'Clocked in successfully',
                    'timestamp': timestamp
                })
            }

        elif action == 'out':
            # Get current session
            try:
                session_response = user_sessions_table.get_item(
                    Key={'user_id': user_id}
                )
                session_item = session_response.get('Item')
                
                if not session_item or session_item.get('status') != 'clocked_in':
                    # Try to clean up any stale clock_in entries from today
                    today = timestamp.split('T')[0]
                    try:
                        # Look for any clock_in entries from today without corresponding clock_out
                        response = time_entries_table.query(
                            KeyConditionExpression=Key('user_id').eq(user_id),
                            FilterExpression=Key('date').eq(today) & Key('type').eq('clock_in')
                        )
                        if response['Items']:
                            # Update session to reflect the found clock_in
                            latest_clock_in = max(response['Items'], key=lambda x: x['timestamp'])
                            user_sessions_table.put_item(
                                Item={
                                    'user_id': user_id,
                                    'status': 'clocked_in',
                                    'clock_in_time': latest_clock_in['timestamp'],
                                    'last_updated': timestamp
                                }
                            )
                            session_item = {
                                'clock_in_time': latest_clock_in['timestamp'],
                                'status': 'clocked_in'
                            }
                        else:
                            return {
                                'statusCode': 400,
                                'headers': headers,
                                'body': json.dumps({'error': 'Not currently clocked in'})
                            }
                    except Exception as cleanup_error:
                        print(f"Session cleanup error: {cleanup_error}")
                        return {
                            'statusCode': 400,
                            'headers': headers,
                            'body': json.dumps({'error': 'Not currently clocked in'})
                        }

                clock_in_time = datetime.fromisoformat(session_item['clock_in_time'].replace('Z', '+00:00'))
                clock_out_time = datetime.fromisoformat(timestamp.replace('Z', '+00:00'))
                hours_worked = Decimal(str(round((clock_out_time - clock_in_time).total_seconds() / 3600, 2)))

                # Create clock out entry
                time_entries_table.put_item(
                    Item={
                        'user_id': user_id,
                        'timestamp': timestamp,
                        'date': date,
                        'type': 'clock_out',
                        'hours': hours_worked,
                        'clock_in_time': session_item['clock_in_time']
                    }
                )

                # Update user session
                user_sessions_table.put_item(
                    Item={
                        'user_id': user_id,
                        'status': 'clocked_out',
                        'last_updated': timestamp
                    }
                )

                return {
                    'statusCode': 200,
                    'headers': headers,
                    'body': json.dumps({
                        'success': True,
                        'message': 'Clocked out successfully',
                        'timestamp': timestamp,
                        'hours': float(hours_worked)
                    })
                }

            except ClientError as e:
                print(f"DynamoDB error: {e}")
                return {
                    'statusCode': 500,
                    'headers': headers,
                    'body': json.dumps({'error': 'Database error'})
                }

        else:
            return {
                'statusCode': 400,
                'headers': headers,
                'body': json.dumps({'error': 'Invalid action. Use "in" or "out"'})
            }

    except json.JSONDecodeError:
        return {
            'statusCode': 400,
            'headers': headers,
            'body': json.dumps({'error': 'Invalid JSON in request body'})
        }
    except Exception as e:
        print(f"Error in clock in/out: {e}")
        return {
            'statusCode': 500,
            'headers': headers,
            'body': json.dumps({'error': 'Internal server error'})
        }

File: Backend/Terraform/lambda_functions/clock_in_out.js
const { connectToDatabase, COLLECTIONS } = require('./utils/database');
const { authenticateUser, createResponse } = require('./utils/auth');
const { validateTimeEntry, formatTimeEntry } = require('./utils/schemas');
const { validateClockAction, sanitizeString } = require('./utils/validation');

exports.handler = async (event) => {
  console.log('Clock in/out request:', JSON.stringify(event, null, 2));

  try {
    // Handle CORS preflight
    if (event.httpMethod === 'OPTIONS') {
      return createResponse(200, { message: 'CORS preflight' });
    }

    // Only allow POST requests
    if (event.httpMethod !== 'POST') {
      return createResponse(405, {
        success: false,
        error: 'Method not allowed'
      });
    }

    // Authenticate user
    const user = await authenticateUser(event);
    
    // Parse request body
    const body = JSON.parse(event.body || '{}');
    const { action, project_id = 'default' } = body;

    // Validate action
    if (!validateClockAction(action)) {
      return createResponse(400, {
        success: false,
        error: 'Invalid action. Must be clock_in or clock_out'
      });
    }

    // Sanitize project_id
    const sanitizedProjectId = sanitizeString(project_id, 100) || 'default';

    // Connect to database
    const { db } = await connectToDatabase();
    
    const now = new Date();
    const dateString = now.toISOString().split('T')[0];

    // Get current user session
    let session = await db.collection(COLLECTIONS.USER_SESSIONS).findOne({
      user_id: user.userId
    });

    if (!session) {
      // Create initial session
      session = {
        user_id: user.userId,
        status: 'out',
        clock_in_time: null,
        project_id: 'default',
        last_updated: now
      };
      await db.collection(COLLECTIONS.USER_SESSIONS).insertOne(session);
    }

    let responseData = {};
    let hours = 0;

    if (action === 'clock_in') {
      // Check if already clocked in
      if (session.status === 'in') {
        return createResponse(400, {
          success: false,
          error: 'Already clocked in'
        });
      }

      // Create clock in time entry
      const clockInEntry = {
        user_id: user.userId,
        timestamp: now,
        date: dateString,
        type: 'clock_in',
        project_id: sanitizedProjectId
      };

      const validation = validateTimeEntry(clockInEntry);
      if (validation.length > 0) {
        return createResponse(400, {
          success: false,
          error: 'Validation failed',
          details: validation
        });
      }

      await db.collection(COLLECTIONS.TIME_ENTRIES).insertOne(clockInEntry);

      // Update user session
      await db.collection(COLLECTIONS.USER_SESSIONS).updateOne(
        { user_id: user.userId },
        {
          $set: {
            status: 'in',
            clock_in_time: now,
            project_id: sanitizedProjectId,
            last_updated: now
          }
        }
      );

      responseData = {
        success: true,
        message: 'Clocked in successfully',
        timestamp: now.toISOString(),
        action: 'clock_in'
      };

    } else if (action === 'clock_out') {
      // Check if clocked in
      if (session.status === 'out') {
        return createResponse(400, {
          success: false,
          error: 'Not currently clocked in'
        });
      }

      // Calculate hours worked
      const clockInTime = new Date(session.clock_in_time);
      hours = (now.getTime() - clockInTime.getTime()) / (1000 * 60 * 60);

      // Create clock out time entry
      const clockOutEntry = {
        user_id: user.userId,
        timestamp: now,
        date: dateString,
        type: 'clock_out',
        project_id: session.project_id,
        hours: parseFloat(hours.toFixed(2)),
        clock_in_time: clockInTime
      };

      const validation = validateTimeEntry(clockOutEntry);
      if (validation.length > 0) {
        return createResponse(400, {
          success: false,
          error: 'Validation failed',
          details: validation
        });
      }

      await db.collection(COLLECTIONS.TIME_ENTRIES).insertOne(clockOutEntry);

      // Update user session
      await db.collection(COLLECTIONS.USER_SESSIONS).updateOne(
        { user_id: user.userId },
        {
          $set: {
            status: 'out',
            clock_in_time: null,
            last_updated: now
          }
        }
      );

      responseData = {
        success: true,
        message: 'Clocked out successfully',
        timestamp: now.toISOString(),
        action: 'clock_out',
        hours: parseFloat(hours.toFixed(2))
      };
    }

    return createResponse(200, responseData);

  } catch (error) {
    console.error('Clock in/out error:', error);
    
    if (error.message === 'Invalid token' || error.message === 'No token provided') {
      return createResponse(401, {
        success: false,
        error: 'Unauthorized'
      });
    }

    return createResponse(500, {
      success: false,
      error: 'Internal server error'
    });
  }
};